pkg/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,82 @@
+package ratelimit
+
+import (
+	"time"
+
+	"go.uber.org/ratelimit"
+)
+
+type Event struct {
+	Timestamp time.Time
+	Tag       int
+}
+
+type RateLimiter struct {
+	outputCh     chan Event
+	src          <-chan Event
+	rt           ratelimit.Limiter
+	ch           chan Event
+	doneSender   chan struct{}
+	doneReceiver chan struct{}
+}
+
+func NewWithEPS(eventsPerSecond int64, src <-chan Event) *RateLimiter {
+	rt := newCommon(src)
+	rt.rt = ratelimit.New(int(eventsPerSecond))
+	return rt
+}
+
+func NewUnlimited(src <-chan Event) *RateLimiter {
+	rt := newCommon(src)
+	rt.rt = ratelimit.NewUnlimited()
+	return rt
+}
+
+func newCommon(src <-chan Event) *RateLimiter {
+	return &RateLimiter{
+		src:          src,
+		ch:           make(chan Event),
+		outputCh:     make(chan Event),
+		doneSender:   make(chan struct{}),
+		doneReceiver: make(chan struct{}),
+	}
+}
+
+func (rt *RateLimiter) Run() {
+
+	sender := func() {
+		for {
+			select {
+			case ev := <-rt.ch:
+				rt.rt.Take()
+				rt.outputCh <- ev
+			case <-rt.doneSender:
+				return
+			}
+		}
+	}
+
+	receiver := func() {
+		for {
+			event := <-rt.src
+			select {
+			case rt.ch <- event:
+			case <-rt.doneReceiver:
+				return
+			default:
+			}
+		}
+	}
+
+	go sender()
+	go receiver()
+}
+
+func (rt *RateLimiter) Stop() {
+	rt.doneReceiver <- struct{}{}
+	rt.doneSender <- struct{}{}
+}
+
+func (rt RateLimiter) OuputChannel() <-chan Event {
```

Comment:
not sure we need a wrapper method: exposing channels as part of the API is a pretty standard practice in golang.

Reply:
You can provide a channel allowing only one directional operation (e.g. only reads, only writes), and I think that's the right approach here https://gobyexample.com/channel-directions
example: https://pkg.go.dev/time#Ticker

Reply:
bah, expressed myself poorly. I'll probably just provide an example to illustrate my point later.

Reply:
do you mean to have a "private" read-write channel and a "public" read-only channel and assign the first one to the second one instead of using a method to do it?

Reply:
type xxx struct {
    outputCh chan Event
    C <-chan Event
}

and eventually do:
C = outputCh